Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,310 @@
+            {
+                return null;
+            }
+            return doc.Value;
+        }
+
+        void IDistributedCache.Refresh(string key)
+        {
+            try
+            {
+                Task.Run(() => _cacheEntries.Document(key)
+                    .UpdateAsync("LastRefresh", _clock.GetCurrentDateTimeUtc())).Wait();
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        async Task IDistributedCache.RefreshAsync(string key, CancellationToken token)
+        {
+            try
+            {
+                await _cacheEntries.Document(key).UpdateAsync(
+                    "LastRefresh", _clock.GetCurrentDateTimeUtc(), cancellationToken: token)
+                    .ConfigureAwait(false);
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        void IDistributedCache.Remove(string key) =>
+            Task.Run(() => _cacheEntries.Document(key).DeleteAsync()).Wait();
+
+        Task IDistributedCache.RemoveAsync(string key, CancellationToken token) =>
+            _cacheEntries.Document(key).DeleteAsync(cancellationToken: token);
+
+        /// <summary>
+        /// Creates a Firestore document that stores the cache value and its
+        /// expiration info, but does not store it in Firestore.
+        /// </summary>
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        private CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = _clock.GetCurrentDateTimeUtc(),
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds = options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = TimeSpan.FromDays(365000);  // 1000 years.
+                doc.SlidingExpirationSeconds = forever.TotalSeconds;
+            }
+            if (options.AbsoluteExpiration.HasValue)
+            {
+                doc.AbsoluteExpiration = options.AbsoluteExpiration.Value.UtcDateTime;
+            }
+            if (options.AbsoluteExpirationRelativeToNow.HasValue)

```

Comment:
If both AbsoluteExpiration and AbsoluteExpirationRelativeToNow are non-null, the latter will win. Is that the intended behavior as far as you're aware? I don't see any documentation on it :(

Reply:
I couldn't find any documentation either.